Add collaboration guideline (AlphaGenes/AlphaPeel#223) (#224)
AlphaGenes/AlphaPeel: docs/source/contribute.rst

Critically, see also a list of issues at `<https://github.com/AlphaGenes/AlphaPeel/issues>`_.

Introducton to collaborative development

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

> Before diving into the technical aspects of contribution, it's essential to understand our collaborative development workflow. 

Reviewer: ```suggestion
Before diving into the technical aspects of contribution,
it's essential to understand our collaborative development workflow. 
```

> This section will outline the rules and best practices for creating issues, pull requests, and branches, as well as our overall workflow from defining tasks to reviewing code.

Reviewer: ```suggestion
This section will outline the rules and best practices for creating issues, branches, and pull requests,
as well as our overall workflow from defining tasks to reviewing code.
```

Issue

> -----

Reviewer: ```suggestion
-----

```

> To efficiently manage our development process, we use tags to categorize issues.

Reviewer: ```suggestion
To efficiently manage our development process,
we use issue tags to categorise issues.
```

> Following are two main tags that we use to label issues:

Reviewer: ```suggestion
Following are two main issue tags that we use to label issues:
```

1. Discussion / Idea

    * Label: ``discussion``

    * No strict format

    * Can stay messy

2. Actionable Task

    * Label: ``task``
    
    * Must follow a template

> **Rule**: Only task issues can have branches/PRs

Reviewer: ```suggestion
**Rule**: Only task issues can have branches/PRs
```

Task issue template
^^^^^^^^^^^^^^^^^^^

> This template is available when you create a new issue at the `AlphaPeel GitHub repository <https://github.com/AlphaGenes/AlphaPeel>`_.

Reviewer: ```suggestion
This template is available when you create a new issue.
```

.. code-block:: markdown

    ## Goal
    What are we trying to achieve?

    ## Proposed approach
    (brief, not perfect)

    ## Scope
    What files/modules are affected?

    ## Done when
    Clear condition for completion

Branch

> ------

Reviewer: ```suggestion
------

```

We use branches to work on issues.

Rules
^^^^^

> * **Branches** are for development, **Tags** are for releases

Reviewer: ```suggestion
* **Branches** are for development, **Version tags** are for releases
```

    * ``devel`` → integration branch

    * ``main`` → release branch

>     * tags only on ``main``

Reviewer: ```suggestion
    * version tags only on ``main``
```

* **Flow**:

    * Feature → PR into ``devel``

>     * Stabilize ``devel``

Reviewer: ```suggestion
    * Stabilise ``devel``
```

    * Merge ``devel`` → ``main``

>     * Tag on ``main``: ``v1.2.0``, ``latest``, ``stable``

Reviewer: ```suggestion
    * Version tag ``main``: ``v1.2.0``, ``latest``, ``stable``
```

Branch naming

> ^^^^^^^^^^^^^

Reviewer: ```suggestion
^^^^^^^^^^^^^

```

A suggested branch naming convention for a task issue on your fork is as follows:

.. code-block:: 

    <type>/issue-<id>-short-description

* Examples:
    * ``feat/issue-42-add-xchrom``
    * ``fix/issue-17-type-error``
    * ``refactor/issue-33-clean-io``
    * ``doc/issue-223-add-collab-guide``

* Types:
    * ``feat``: new functionality
    * ``fix``: bug fix
    * ``refactor``: code cleanup
    * ``doc``: documentation updates

Pull request

> ------------

Reviewer: ```suggestion
------------

```

A pull request (PR) is a request to merge your code changes from your branch into the main repository's branch (e.g., ``devel``).

> The followings are the templates for PRs of AlphaPeel and tinyhouse, which the latter is the submodule referenced by the former. 

Reviewer: ```suggestion
The followings are the templates for PRs of ``AlphaPeel`` and ``tinyhouse``,
which the latter is the submodule referenced by the former. 
```

> The PR template should be followed when you create a PR to the AlphaPeel repository or the tinyhouse repository.

Reviewer: ```suggestion
The PR template should be followed when you create a PR to either
the ``AlphaPeel`` repository or the ``tinyhouse`` repository.
```

> We encourage you to open PRs early, even when the code is not fully ready, to facilitate early feedback and discussion. You can use draft PRs to indicate that the PR is a work in progress.

Reviewer: ```suggestion
We encourage you to open PRs early,
even when the code is not fully ready,
to facilitate early feedback and discussion.
You can use draft PRs to indicate that the PR is a work in progress.
```

Template for PR of AlphaPeel

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
Template for PR of ``AlphaPeel``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

.. code-block:: markdown

    ## Related Issue
    Closes #<issue-id>

    ## What changed
    - Brief summary of changes
    - Key files/modules affected

    ## Why this change
    - What problem does this solve?
    - What functionality does it enable?

    ## Submodule changes
    - Submodule repo: <name>
    - PR: <link> (if applicable)
    - Commit: <hash>
    - Why needed: <what functionality depends on this>

    ## Notes / Risks
    - Anything reviewers should pay attention to

Template for PR of tinyhouse

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
Template for PR of ``tinyhouse``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

.. code-block:: markdown

    ## Related Issue
    <submodule-issue-id or main-repo issue link>

    ## What changed
    - Summary of changes

    ## Why this change
    - What functionality this enables

    ## Used by
    - Main repo PR: <link> (if applicable)

    ## Notes
    - Any compatibility considerations

Full workflow
-------------

> Below is the full workflow for contribution and development, from defining the work to merging a pull request.

Reviewer: ```suggestion
Below is the full workflow for contribution and development,
from defining the work to merging a pull request.
```

1️⃣ Create and define the work
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

>     In AlphaPeel repo:

Reviewer: ```suggestion
    In ``AlphaPeel`` repo:
```

    * Create a **task issue**
    * Clearly define:

        * goal
        * scope
        * “done when”

>     If tinyhouse changes are needed:

Reviewer: ```suggestion
    If ``tinyhouse`` changes are needed:
```

>     * create a **linked issue in tinyhouse**

Reviewer: ```suggestion
    * create a **linked issue in ``tinyhouse``**
```

2️⃣ Create branches (both repos)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    In your fork:

>     * AlphaPeel repo example:

Reviewer: ```suggestion
    * ``AlphaPeel`` repo example:
```

        .. code-block:: bash

            feat/issue-42-new-analysis
        

>     * tinyhouse repo example:

Reviewer: ```suggestion
    * ``tinyhouse`` repo example:
```

        .. code-block:: bash

            feat/issue-15-support-analysis

3️⃣ Implement tinyhouse changes first

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
3️⃣ Implement ``tinyhouse`` changes first
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

    Steps:

>     * Implement changes in tinyhouse branch

Reviewer: ```suggestion
    * Implement changes in ``tinyhouse`` branch
```

    * Push to your fork

>     * Open **PR to devel branch of tinyhouse repo**

Reviewer: ```suggestion
    * Open **PR to ``devel`` branch of ``tinyhouse`` repo**
```

4️⃣ Use updated tinyhouse reference in AlphaPeel repo

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
4️⃣ Use updated ``tinyhouse`` reference in ``AlphaPeel`` repo
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

>     In your AlphaPeel repo branch:

Reviewer: ```suggestion
    In your ``AlphaPeel`` repository branch:
```

>         * Point tinyhouse submodule to your **branch commit**

Reviewer: ```suggestion
        * Point ``tinyhouse`` submodule to your **branch commit**
```

>         * Continue development and testing the changes in AlphaPeel repo

Reviewer: ```suggestion
        * Continue development and testing the changes in your ``AlphaPeel`` repository
```

5️⃣ Open PRs early
^^^^^^^^^^^^^^^^^^^^^^

    Open:

>     * tinyhouse PR → target: tinyhouse ``AlphaGenes:devel``

Reviewer: ```suggestion
    * ``tinyhouse`` PR → target: ``AlphaGenes:tinyhouse/devel``
```

Author: I got the idea. Usually on GitHub, the colon and the slash are the other way around, such as ``AlphaGenes/AlphaPeel:main``. I will do it this way.

>     * AlphaPeel repo PR → target: AlphaPeel ``AlphaGenes:devel``

Reviewer: ```suggestion
    * ``AlphaPeel`` repo PR → target:``AlphaGenes:AlphaPeel/devel``
```

    Link them both ways:

>     * AlphaPeel repo PR → links tinyhouse PR

Reviewer: ```suggestion
    * ``AlphaPeel`` repo PR → links ``tinyhouse`` PR
```

>     * tinyhouse PR → links AlphaPeel repo PR

Reviewer: ```suggestion
    * ``tinyhouse`` PR → links ``AlphaPeel`` repo PR
```

6️⃣ Development + syncing

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
6️⃣ Development + syncing
^^^^^^^^^^^^^^^^^^^^^^^^
```

    During development, always keep your branch synced with upstream using ``rebase``.

7️⃣ Review phase

> ^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
7️⃣ Review phase
^^^^^^^^^^^^^^^
```

>     * First: tinyhouse PR

Reviewer: ```suggestion
    * First: ``tinyhouse`` PR
```

>         * Review and merge tinyhouse PR

Reviewer: ```suggestion
        * Review and merge ``tinyhouse`` PR
```

        * Now you have a **stable commit hash**

>     * Then: Update AlphaPeel repo

Reviewer: ```suggestion
    * Then: Update ``AlphaPeel`` repo
```

>         * Update tinyhouse submodule pointer to the **merged commit**

Reviewer: ```suggestion
        * Update ``tinyhouse`` submodule pointer to the **merged commit**
```

>         * Push update to AlphaPeel repo branch

Reviewer: ```suggestion
        * Push update to ``AlphaPeel`` repo branch
```

>     * Then: AlphaPeel repo PR review

Reviewer: ```suggestion
    * Then: ``AlphaPeel`` repositoryp PR review
```

8️⃣ Merge flow

> ^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
8️⃣ Merge flow
^^^^^^^^^^^^^
```

    * Before merge: 
        * final check the new changes by others and update with ``rebase`` if needed
        * ``squash`` the commits to simplify the history

Weekly meetings

> ---------------

Reviewer: ```suggestion
Developer meetings
------------------

```

> We have weekly meetings to discuss the ongoing work and plan for the next steps.

Reviewer: ```suggestion
We have routine developer meetings to discuss the ongoing work and plan for the next steps.
```
